### PR TITLE
[persist] Clean up some disused filter-pushdown code 

### PR DESCRIPTION
**Pull request by bkirwi**
Removes some now-dead filter-pushdown-related code.

### Motivation

This PR refactors existing code. Deleting unused code is generally considered wise.

In support of: #18007.

### Tips for reviewer

Most of the green lines in the diff are the one test I had to move to avoid making a bunch of methods unnecessarily public. It's all code movement: I didn't have to make any changes aside from some imports.

This starts to bake in the idea that `ResultSpec` is the boundary between `expr` and `persist`; note the dependency drop. This seems overall positive so far, but worth calling out.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by danhhz — APPROVED** [review]
LGTM!

Making the interpreter the API boundary between compute+storage instead of the trait you're removing here sounds great to me, but it's a conceptual shift from what we agreed with compute, so we should make sure to get signoff from them. (Plus we should update the section in the design doc declaring what the boundary is, but I don't think making that update needs to block merging this PR if we get consensus.)

**Comment by bkirwi** [comment]
Cool!

@MaterializeInc/compute - the change here is that while originally we'd added a dependency from `expr` to `persist-types` so the MFP-pushdown code could read statistics, it's now clear that the persist source can live entirely downstream of the new `Interpreter`, providing statistics to the interpreter by constructing a `ResultSpec` directly, and we can remove the dependency. I think this is only upside for y'all - fewer dependencies in the crates you're responsible for - but I'll hold off merging this for a couple days in case there are any concerns on your end!

File: src/expr/src/linear.rs
Comment (aalexandrov) [inline src/expr/src/linear.rs:1890]: Why is this no longer needed?

**Comment by aalexandrov** [comment]
> it's now clear that the persist source can live entirely downstream of the new Interpreter

In general it would be great if we can have some decoupling between:

1. the interface definition (`Interpreter`), and 
1. the implementations that provide the module level facilities:
    1. all `Interpreter` definitions and 
    1. the associated `Interpreter::Summary` types), and.

* At the very least (1) and (2) should live in different modules under `mz_expr`.
* If it is not problematic to move (2) under `storage-client` I think this would be even better, as at the moment it seems to be the only consumer of the `Interpreter` implementations. If we figure out we need more consumers we can promote those back under `mz_expr`.

**Review by aalexandrov — APPROVED** [review]
Overall LGTM, I left some black spots and more general ideas of cleaner code separation, but those can be addressed later if needed.

File: src/expr/src/linear.rs
Comment (bkirwi) [reply src/expr/src/linear.rs:1890]: This code was effectively doing a hard-coded subset of the abstract-interpretation problem that is now solved in general! There's still a little code that translates between persist stats and `ResultSpec`s, which now lives in the Persist source.

**Comment by bkirwi** [comment]
Thanks for the review!

Thanks also for the comments on code structure, though I am tempted to leave those for a followup since the changes here are approved.

I'm generally happy to structure modules any way that harmonizes with the rest of `expr`. I think the case for moving it into `storage-client` is a bit harder; I think it's benefitted a lot from being treated as a general thing and not a storage detail, and y'all on Compute have a ton of relevant expertise. But I'm open to it especially if it proves burdensome...

**Comment by aalexandrov** [comment]
> I think it's benefitted a lot from being treated as a general thing and not a storage detail, and y'all on Compute have a ton of relevant expertise.

Moving the implementations to `storage-client` does not mean that we shouldn't review changes to these implementations! If this is a concern we can just put @MaterializeInc/compute as a shared owner of the specific `storage-client` module that hosts them!

**Pull request merged**
